### ISSUE TITLE
EasyOCR Finetuning Error in dataloder

### ISSUE THREAD

**Issue by nadhirhasan** [JaidedAI/EasyOCR#1002]
Hello,

I have been trying to train a custom model using the trainer folder in this particular repository. However, I have encountered several errors due to version updates. After much debugging, I have encountered a strange error related to the use of a data loader. Specifically, I am receiving the error message: AttributeError: '_MultiProcessingDataLoaderIter' object has no attribute 'next'.

I have tried using both data_loader_iter.next() and next(data_loader_iter) to load the data, but the error persists. I am hoping someone can provide me with some assistance in resolving this issue. Thank you.

**Comment by s-shakir** [comment]
I faced this similar issue, trying to train custom model. I made the following changes and it works fine for me now:

data_loader_iter= iter(data_loader_iter)
data_loader_iter= next(data_loader_iter)

its because of the difference in the pytorch versions

**Comment by nadhirhasan** [comment]
Thank you for your response, however, I am still encountering the same error. I would like to share my code on a Kaggle notebook and I was wondering if you could review it when you have some free time. It's possible that the problem lies on my side.

**Here's the link to my notebook**: https://www.kaggle.com/nadhirhasan/easyocr-finetune

This is what tried

`class Batch_Balanced_Dataset(object):

    def __init__(self, opt):
        """
        Modulate the data ratio in the batch.
        For example, when select_data is "MJ-ST" and batch_ratio is "0.5-0.5",
        the 50% of the batch is filled with MJ and the other 50% of the batch is filled with ST.
        """
        log = open(f'./saved_models/{opt.experiment_name}/log_dataset.txt', 'a')
        dashed_line = '-' * 80
        print(dashed_line)
        log.write(dashed_line + '\n')
        print(f'dataset_root: {opt.train_data}\nopt.select_data: {opt.select_data}\nopt.batch_ratio: {opt.batch_ratio}')
        log.write(f'dataset_root: {opt.train_data}\nopt.select_data: {opt.select_data}\nopt.batch_ratio: {opt.batch_ratio}\n')
        assert len(opt.select_data) == len(opt.batch_ratio)

        _AlignCollate = AlignCollate(imgH=opt.imgH, imgW=opt.imgW, keep_ratio_with_pad=opt.PAD, contrast_adjust = opt.contrast_adjust)
        self.data_loader_list = []
        self.dataloader_iter_list = []
        batch_size_list = []
        Total_batch_size = 0
        for selected_d, batch_ratio_d in zip(opt.select_data, opt.batch_ratio):
            _batch_size = max(round(opt.batch_size * float(batch_ratio_d)), 1)
            print(dashed_line)
            log.write(dashed_line + '\n')
            _dataset, _dataset_log = hierarchical_dataset(root=opt.train_data, opt=opt, select_data=[selected_d])
            total_number_dataset = len(_dataset)
            log.write(_dataset_log)

            """
            The total number of data can be modified with opt.total_data_usage_ratio.
            ex) opt.total_data_usage_ratio = 1 indicates 100% usage, and 0.2 indicates 20% usage.
            See 4.2 section in our paper.
            """
            number_dataset = int(total_number_dataset * float(opt.total_data_usage_ratio))
            dataset_split = [number_dataset, total_number_dataset - number_dataset]
            indices = range(total_number_dataset)
            _dataset, _ = [Subset(_dataset, indices[offset - length:offset])
                           for offset, length in zip(_accumulate(dataset_split), dataset_split)]
            selected_d_log = f'num total samples of {selected_d}: {total_number_dataset} x {opt.total_data_usage_ratio} (total_data_usage_ratio) = {len(_dataset)}\n'
            selected_d_log += f'num samples of {selected_d} per batch: {opt.batch_size} x {float(batch_ratio_d)} (batch_ratio) = {_batch_size}'
            print(selected_d_log)
            log.write(selected_d_log + '\n')
            batch_size_list.append(str(_batch_size))
            Total_batch_size += _batch_size

            _data_loader = torch.utils.data.DataLoader(
                _dataset, batch_size=_batch_size,
                shuffle=True,
                num_workers=int(opt.workers), #prefetch_factor=2,persistent_workers=True,
                collate_fn=_AlignCollate, pin_memory=True)
            self.data_loader_list.append(_data_loader)
            self.dataloader_iter_list.append(iter(_data_loader))

        Total_batch_size_log = f'{dashed_line}\n'
        batch_size_sum = '+'.join(batch_size_list)
        Total_batch_size_log += f'Total_batch_size: {batch_size_sum} = {Total_batch_size}\n'
        Total_batch_size_log += f'{dashed_line}'
        opt.batch_size = Total_batch_size

        print(Total_batch_size_log)
        log.write(Total_batch_size_log + '\n')
        log.close()

    def get_batch(self):
        balanced_batch_images = []
        balanced_batch_texts = []

        for i, data_loader_iter in enumerate(self.dataloader_iter_list):
            try:
                image,text = next(iter(data_loader_iter))
                balanced_batch_images.append(image)
                balanced_batch_texts += text
            except StopIteration:
                self.dataloader_iter_list[i] = iter(self.data_loader_list[i])
                image, text = iter(self.dataloader_iter_list[i])
                balanced_batch_images.append(image)
                balanced_batch_texts += text
            except ValueError:
                pass

        balanced_batch_images = torch.cat(balanced_batch_images, 0)

        return balanced_batch_images, balanced_batch_texts
`

Error:
`AttributeError                            Traceback (most recent call last)
/tmp/ipykernel_23/1680820862.py in <module>
----> 1 train(opt, amp=False)

/kaggle/working/train.py in train(opt, show_number, amp)
    201             scaler.step(optimizer)
    202             scaler.update()
--> 203         else:
    204             image_tensors, labels = train_dataset.get_batch()
    205             image = image_tensors.to(device)

/kaggle/working/dataset.py in get_batch(self)
     99         balanced_batch_texts = []
    100 
--> 101         for i, data_loader_iter in enumerate(self.dataloader_iter_list):
    102             try:
    103                 image,text = next(iter(data_loader_iter))

AttributeError: '_MultiProcessingDataLoaderIter' object has no attribute 'next'`

**Comment by nadhirhasan** [comment]
> I faced this similar issue, trying to train custom model. I made the following changes and it works fine for me now:
> 
> data_loader_iter= iter(data_loader_iter) data_loader_iter= next(data_loader_iter)
> 
> its because of the difference in the pytorch versions

thanks mam, it's working

**Comment by Srini138** [comment]
> > I faced this similar issue, trying to train custom model. I made the following changes and it works fine for me now:
> > data_loader_iter= iter(data_loader_iter) data_loader_iter= next(data_loader_iter)
> > its because of the difference in the pytorch versions
> 
> thanks mam, it's working

@nadhirhasan  Hi, could you please tell where to change these lines in dataset.py

**Comment by masoudMZB** [comment]
@Srini138 inside dataset.py ->   def get_batch(self) 
there is a for loop, and inside for there is a try except. put it there

**Comment by nadhirhasan** [comment]
> @Srini138 inside dataset.py -> def get_batch(self) there is a for loop, and inside for there is a try except. put it there

you can find it on `Batch_Balanced_Dataset`. from a function called `get_batch`

**Comment by nadhirhasan** [comment]
> where



> > > I faced this similar issue, trying to train custom model. I made the following changes and it works fine for me now:
> > > data_loader_iter= iter(data_loader_iter) data_loader_iter= next(data_loader_iter)
> > > its because of the difference in the pytorch versions
> > 
> > 
> > thanks mam, it's working
> 
> @nadhirhasan Hi, could you please tell where to change these lines in dataset.py

yeah, @masoudMZB pointing the correct place. you can found within the get_batch function of the Batch_Balanced_Dataset class.

**Comment by harivinod3** [comment]
Hello,

I too got the same error and I tried to place the given code in the dataset.py. But its not working. Can you please send that function code where you changed it. Thank you

**Comment by nadhirhasan** [comment]
hi @harivinod3,

Perhaps your code is correct, but there might be issues with the data format and file structure. I've recently created a notebook on finetuning EasyOCR, which I believe can provide you with a better understanding if you take a look.

EasyOCR finetuning: https://www.kaggle.com/code/nadhirhasan/let-s-finetune-easyocr-boom/notebook
Inference Notebook: https://www.kaggle.com/code/nadhirhasan/check-the-performance-of-our-finetuned-ocr-model
